Checked inverse

markscheel opened the pull request:
## Proposed changes

The `inverse` function of all CoordinateMaps now returns `boost::optional<...>`, which holds an invalid value if one tries to inverse-map a point that is sufficiently outside the range of the map, especially a point for which the map is singular or not one-to-one.

Instantiations of inverse for DataVectors have been removed, since we don't see a use case (outside of tests) for inverse acting on a DataVector.  Similarly, ElementMap::inverse is no longer used/tested for DataVectors.

This is a breaking change for any code that uses inverse maps.

This is a large PR because it needs to touch all the maps, all the tests for them, and any place that the inverse map is used (the latter is not many).

### Types of changes:

- [ ] Bugfix
- [x] New feature

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests, `clang-tidy` and `IWYU`. For
  instructions on how to perform the CI checks locally refer to the [Dev guide
  on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run `make doc`
  to generate the documentation locally into `BUILD_DIR/docs/html`. Then open
  `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

markscheel committed:
All CoordMaps now have checked inverse functions.

inverse now returns boost::optional<>, which can be invalid
if one tries to inverse-map a point that is sufficiently outside
the range of the map, especially a point for which the map is
singular or not one-to-one.

Instantiations of inverse for DataVectors have been removed, since we
don't see a use case (outside of tests) for inverse acting on a DataVector.

Similarly, ElementMap::inverse is no longer used/tested for DataVectors.

Also, instantiations of inverse for std::reference_wrapper<const double>
have been removed, since we do not need these either.

markscheel commented:
How do you get codecov to rerun?  The current diff report shows diffs from older versions of the code.

markscheel commented:
(I just rebased because of a conflict with recently merged PRs)

markscheel commented:
The failing tests are clang_tidy complaining about code in boost/toms748

mar-celine commented on src/Domain/CoordinateMaps/BulgedCube.cpp:
```diff
@@ -37,7 +37,7 @@ class RootFunction {
 
   double operator()(const double rho, const size_t i = 0) const noexcept {
-    if (not(equal_within_roundoff(get_element(physical_r_squared_, i), 0.0))) {
+    if (LIKELY(get_element(physical_r_squared_, i) != 0.0)) {
```
The rootfind is only done for the inverse (which only works on doubles) so we don't need this to be templated on DType anymore. Also since get_element was made to be used on both doubles and DataVectors we won't need to use get_element once this is hard-coded on doubles only

mar-celine commented on src/Domain/CoordinateMaps/CubicScale.hpp:
```diff
@@ -43,8 +44,10 @@ class CubicScale {
       const std::unordered_map<std::string, FunctionOfTime&>& map_list) const
       noexcept;
 
+  /// Returns invalid if point is out of the range of the map, or if the
```
I think `Returns boost::none` is more concrete than `Returns invalid`

mar-celine commented on src/Domain/CoordinateMaps/EquatorialCompression.cpp:
```diff
@@ -37,12 +37,21 @@ EquatorialCompression::angular_distortion(const std::array<T, 3>& coords,
-      ASSERT(equal_within_roundoff(get_element(radius_over_rho, i), 0.0),
-             "r == 0 must imply rho == 0. This has failed.");
     }
+    // There is no 'else' covering the case rho==0.  The only way that
```
Thank you for this extra detail :+1:

mar-celine commented on src/Domain/CoordinateMaps/Frustum.hpp:
```diff
@@ -51,8 +52,11 @@ class Frustum {
       const std::array<T, 3>& source_coords) const noexcept;
 
+  /// Returns invalid if \f$z\f$ is at or beyond the \f$z\f$-coordinate of
+  /// the apex of the pyramid formed by extending the `Frustum` (for a
```
Extended Frustums only end a pyramid if one of the bases of the frustum is smaller in both directions than the other base. In cases where the bases are the same length along one direction, the extended frustum makes a triangular prism. A third case where one base is longer in one direction and shorter in the second direction leads to a tetrahedron instead. 

But I think this can be reworded such that it works for all cases, since these are all triangles in the x-z and y-z planes, and the apex of the triangle in the plane is what will determine where the mapping becomes singular.

mar-celine commented on src/Domain/CoordinateMaps/Wedge3D.cpp:
```diff
@@ -109,14 +111,32 @@ std::array<tt::remove_cvref_wrap_t<T>, 3> Wedge3D::inverse(
-  const auto zeta_coefficient =
+  const ReturnType zeta_coefficient =
       (scaled_frustum_rate_ + sphere_rate_ * one_over_rho);
+  // If -sphere_rate_/scaled_frustum_rate_ > 1, then
```
Might be easier to parse if written as `If -sphere_rate_ > scaled_frustum_rate_`

markscheel replied:
That doesn't work:  `-a/b > 1`  is equivalent to `-a>b` only if `b` is positive; I think the comment as written works for all cases.

mar-celine commented on src/Domain/CoordinateMaps/Wedge3D.hpp:
```diff
@@ -237,8 +238,16 @@ class Wedge3D {
 
+  /// For a \f$+z\f$-oriented `Wedge3D`, returns invalid if \f$z<=0\f$
+  /// or if \f$(x,y,z)\f$ is on or outside the paraboloid defined
+  /// by \f$(x^2/z^2 + y^2/z^2+1)^{1/2} = -S/F\f$, where
```
I might have made a mistake, but I'm getting that this is a cone shape?

markscheel replied:
Yes!  It's a cone, sorry, will fix.

mar-celine commented on tests/Unit/Domain/CoordinateMaps/Test_ProductMaps.cpp:
```diff
@@ -4,18 +4,58 @@
+    // Should be ok
+    const std::array<double, 3> mapped_point2{{3.5, 3.0, -1.0}};
+
+    CHECK(not map.inverse(mapped_point1));
```
Why not CHECK_FALSE and CHECK (to avoid `not`s?)

markscheel replied:
The `not s` casts to a bool, which is why we want it.  Equivalently we could have `static_cast<bool>(s)` instead.  (If you don't cast to bool, you get a compiler error about boost::optional not having stream operators.  I don't think we care about writing stream operators for boost::optional just for a test where all you care about is whether the boost::optional is true or false).  I'm ok with changing to static_cast if people like it better.  Thoughts?

markscheel replied:
I'm changing to `static_cast<bool>` because that makes the logic clearer and explicitly says what I am doing.

mar-celine reviewed (commented):
I did not check CubicScale as closely, maybe @mgiesler can take a look at it?

mgiesler commented:
Per @gsb76's request, I took a quick look at CubicScale and it looks good to me :+1:

wthrowe commented on src/Domain/CoordinateMaps/CubicScale.cpp:
```diff
@@ -50,18 +48,19 @@ std::array<tt::remove_cvref_wrap_t<T>, 1> CubicScale::inverse(
-          << " does not satisfy this criterion.");
+  if (a_of_t <= 0.0 or b_of_t < 2.0 / 3.0 * a_of_t or outer_boundary_ == 0.0 or
+      b_of_t == 0.0) {
+    return boost::none;
```
I think these are testing that the map itself is not invalid.  If so, I think we should leave them as errors, since if you even get the map into this state something has gone wrong.

wthrowe commented on src/Domain/CoordinateMaps/Frustum.cpp:
```diff
@@ -105,12 +107,20 @@ std::array<tt::remove_cvref_wrap_t<T>, 3> Frustum::inverse(
   // We reuse `coords` to avoid an allocation. This is now the
   // logical coords {xi,eta,zeta}
```
Since we now know ReturnType (and similar in the other maps) is a double we should probably
a) get rid of it and used `double` explicitly, and
b) get rid of all the tricks that make things more confusing in the name of eliminating allocations (since there are no allocations anymore).

markscheel replied:
Do we still want inverse() to be able to take as an argument a `const std::array<std::reference_wrapper<const double>,3>` as well as a `const std::array<double,3>`?

nilsdeppe replied:
If I understood our conversation yesterday, the consensus was no ref_wraps in inverse (at least for now)

wthrowe commented on tests/Unit/Domain/CoordinateMaps/Test_Frustum.cpp:
```diff
@@ -48,8 +49,35 @@ void test_suite_for_frustum(const bool with_equiangular_map) {
+  const std::array<double, 3> test_mapped_point3{{6.0, -7.0, -3.0}};
+
+  // Should work, even though it is outside the mapped frustum.
+  const std::array<double, 3> test_mapped_point4{{0.0, 0.0, 9.0}};
```
Why do we want to guarantee this works?  People should not be relying on the ability to reverse-map points that are out of the range of the map.

wthrowe commented on tests/Unit/Domain/CoordinateMaps/Test_Frustum.cpp:
```diff
@@ -48,8 +49,35 @@ void test_suite_for_frustum(const bool with_equiangular_map) {
+  CHECK(not map.inverse(test_mapped_point1));
+  CHECK(not map.inverse(test_mapped_point2));
+  CHECK(not map.inverse(test_mapped_point3));
+  CHECK_FALSE(not map.inverse(test_mapped_point4));
```
Cast to bool instead of double negative.

wthrowe commented on tests/Unit/Domain/CoordinateMaps/Test_Wedge2D.cpp:
```diff
@@ -123,8 +124,32 @@ void test_wedge2d_all_orientations(const bool with_equiangular_map) {
+  const std::array<double, 2> test_mapped_point3{{-1.0, 3.0}};
+
+  // This point should be fine, even though it is outside the mapped wedge.
+  const std::array<double, 2> test_mapped_point4{{100.0, -6.0}};
```
Same comments as for frustum.

wthrowe commented on tests/Unit/Domain/CoordinateMaps/Test_Wedge3D.cpp:
```diff
@@ -233,8 +234,43 @@ void test_wedge3d_random_radii(const bool with_equiangular_map) {
         approx(random_outer_radius_upper_zeta));
 }
+
+void test_wedge3d_fail() noexcept {
```
Same comments as for Frustum.

markscheel committed:
Instantiate Identity<3>.

A comment in Identity.cpp said that Identity<3> shouldn't be instantiated,
but it is used e.g. in Test_InitializedElement.cpp. Furthermore,
member functions of Identity<3> are explicitly instantiated.

markscheel commented:
Rebased and squashed. But I left one commit separate (instantiating `Identity<3>`) because someone should look at that commit (and my commit message for it), since it fixes a contradiction between a comment and actual usage of `Identity<3>`.

nilsdeppe commented:
@wthrowe have your concerns been addressed?

markscheel commented:
I just fixed lingering IWYU and clang-tidy problems.

markscheel commented:
I did another IWYU change.  The remaining clang-tidy warnings are in toms748 (part of boost).

nilsdeppe reviewed (dismissed):
@wthrowe have your concerns been addressed?

wthrowe commented on src/Domain/CoordinateMaps/CubicScale.hpp:
```diff
@@ -43,8 +44,10 @@ class CubicScale {
       noexcept;
 
+  /// Returns boost::none if point is out of the range of the map, or if the
+  /// map is not invertible.
```
This comment needs updating.

wthrowe commented on tests/Unit/Domain/CoordinateMaps/Test_BulgedCube.cpp:
```diff
@@ -48,6 +48,33 @@ SPECTRE_TEST_CASE("Unit.Domain.CoordinateMaps.BulgedCube.Identity",
 
+void test_bulged_cube_fail() {
+  const std::array<double, 3> lower_corner{{-1.0, -1.0, -1.0}};
+  const std::array<double, 3> upper_corner{{1.0, 1.0, 1.0}};
```
These are never used.

wthrowe commented on tests/Unit/Domain/CoordinateMaps/Test_BulgedCube.cpp:
```diff
@@ -48,6 +48,33 @@ SPECTRE_TEST_CASE("Unit.Domain.CoordinateMaps.BulgedCube.Identity",
+  const std::array<double, 3> test_mapped_point3{{4.0, 0.0, 0.0}};
+
+  // These points work
+  const std::array<double, 3> test_mapped_point4{{2.0, 2.0, 2.0}};
```
I think this is another case of testing things we don't want to guarantee.

wthrowe commented on tests/Unit/Domain/CoordinateMaps/Test_BulgedCube.cpp:
```diff
@@ -48,6 +48,33 @@ SPECTRE_TEST_CASE("Unit.Domain.CoordinateMaps.BulgedCube.Identity",
+  CHECK(not map.inverse(test_mapped_point2));
+  CHECK(not map.inverse(test_mapped_point3));
+  CHECK_FALSE(not map.inverse(test_mapped_point4));
+  CHECK_FALSE(not map.inverse(test_mapped_point5));
```
Cast to bool.

wthrowe commented on tests/Unit/Domain/CoordinateMaps/Test_CubicScale.cpp:
```diff
@@ -192,74 +235,39 @@ SPECTRE_TEST_CASE(
+    "[Domain][Unit]") {
+  ERROR_TEST();
+  // Make b<0
+  cubic_scale_non_invertible(0.96, 0.0, 20.0);
```
Comment does not match code.

kidder reviewed (commented):
I am happy with this PR.  @wthrowe please look at

wthrowe reviewed (dismissed):
Looks good.  Squash.

markscheel committed:
TestMapHelpers: rename a helper function.

markscheel committed:
TestMapHelpers: add test_suite_for_map_on_sphere.

nilsdeppe reviewed (dismissed):
@gsb76 @wthrowe @kidder whenever you have time, mark added the spherical test function we discussed today.

markscheel commented:
I just fixed an IWYU problem, and a roundoff problem in SpecialMobius::inverse that caused a test to fail every so often.

mar-celine commented on tests/Unit/Domain/CoordinateMaps/TestMapHelpers.hpp:
```diff
@@ -338,6 +338,71 @@ void test_suite_for_map_on_unit_cube(const Map& map) {
 
+/*!
+ * \ingroup TestingFrameworkGroup
+ * \brief Given a Map `map`, tests the map functions, including map inverse,
```
document or rename `max_radius` such that it is clear that it the radius used for the boundary of the map and that the test expects that the map still be valid on the boundary.

You also might want to differentiate the documentation for this test from that for test_suite_for_map_on_unit_cube, also because this test only works for 3D and only the code mentions that

markscheel commented:
Improved documentation for test_suite_for_map_on_sphere

wthrowe commented on tests/Unit/Domain/CoordinateMaps/TestMapHelpers.hpp:
```diff
@@ -338,6 +340,77 @@ void test_suite_for_map_on_unit_cube(const Map& map) {
+  const auto seed = std::random_device{}();
+  std::mt19937 gen(seed);
+  INFO("seed = " << seed);
+  std::uniform_real_distribution<> radius_dis(0.0, radius_of_sphere);
```
Probably should stay away from the origin in the include_origin=false case.

nilsdeppe replied:
What value should the interior bound be in that case? `numeric_limits::epsilon * radius_of_sphere`?

wthrowe replied:
`test_jacobian` uses a finite difference stencil of half-width 3e-4.  If the map can't be evaluated at the origin it probably isn't numerically well-behaved near there, so we probably want to keep this stencil well away from r=0.  I wouldn't go less than 1e-3, probably even larger.

In any case, it might be a good idea to test a random point on this inner boundary, (similar to what is done for the outer boundary) not because the surface is intrinsically interesting, but because it would flush out RNG-induced test failures quicker.

markscheel commented:
There is now an `inner_bdry` set to 5e-3 for the case `include_origin == false`, and a random point on this boundary is tested.  (of course we don't yet have a map where we set `include_origin==false`)

wthrowe commented:
Needs a rebase.

markscheel commented:
Had to rebase, and change one thing in BlockLogicalCoords where inverse was used.

nilsdeppe commented:
travis failures are timeouts, all tests passed.
